src/libraries/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -1753,38 +1753,38 @@ bool ISpanFormattable.TryFormat(Span<char> destination, out int charsWritten, Re
             // be necessary for semantics inside of TryFormatPrimitiveNonDefault, so we can just do it here instead.
             if (format.IsEmpty)
             {
-                if (underlyingType == typeof(int)) return TryFormatPrimitiveDefault(rt, *(int*)&value, destination, out charsWritten);
-                if (underlyingType == typeof(uint)) return TryFormatPrimitiveDefault(rt, *(uint*)&value, destination, out charsWritten);
-                if (underlyingType == typeof(long)) return TryFormatPrimitiveDefault(rt, *(long*)&value, destination, out charsWritten);
-                if (underlyingType == typeof(ulong)) return TryFormatPrimitiveDefault(rt, *(ulong*)&value, destination, out charsWritten);
-                if (underlyingType == typeof(byte)) return TryFormatPrimitiveDefault(rt, *(byte*)&value, destination, out charsWritten);
-                if (underlyingType == typeof(sbyte)) return TryFormatPrimitiveDefault(rt, *(sbyte*)&value, destination, out charsWritten);
-                if (underlyingType == typeof(short)) return TryFormatPrimitiveDefault(rt, *(short*)&value, destination, out charsWritten);
-                if (underlyingType == typeof(ushort)) return TryFormatPrimitiveDefault(rt, *(ushort*)&value, destination, out charsWritten);
+                if (underlyingType == typeof(int)) return TryFormatPrimitiveDefault(rt, Unsafe.BitCast<TEnum, int>(value), destination, out charsWritten);
```

Comment:
This is going to add a ton of new generic instantiations that is going to have a measurable footprint impact of enum formatting.

Does this improve throughput of enum formatting in a measurable way?

Reply:
(More places in this PR have same problem. I am not commenting on all of them.)

Reply:
> Does this improve throughput of enum formatting in a measurable way?

Looks like "no", based on the SPMI report, although since that notably doesn't include VM calls I'll try to get more direct numbers as well.

What's required to remove the generic instantiations from these types of intrinsics where they aren't "actually" needed. That seems like a good long term thing to remove given how prevalent generics are, particularly for JIT intrinsics and simple methods which are effectively "always inlined".

Reply:
> What's required to remove the generic instantiations from these types of intrinsics where they aren't "actually" needed. 

I think we would always need something to represent these generic instantiations. If we do not want this representation to persist, we would have to decouple the run-time and compile-time type systems. The run-time type system would be permanent like it is today. The compile-time type system would be transient. It would get flushed every once in a while.

It would be a major redesign of the JIT/EE interface implementation.

Reply:
> I think we would always need something to represent these generic instantiations. If we do not want this representation to persist, we would have to decouple the run-time and compile-time type systems.

Not sure I quite understand in this context.

What we have is `Unsafe.BitCast<T, U>(value)` in IL. As we're processing the callee, we end up resolving the call token and then calling `eeGetCallInfo` which gets the information on the call populating the `CORINFO_CALL_INFO*`. Doing this requires some instantiations on the VM side to populate the relevant information.

However, in this particular case we have a known JIT intrinsic and since its a special one like `System.Runtime.CompilerServices.Unsafe`, we can also know that its either going to be completely handled up front with no preservation of the generic instantiation for the method or we'll fail it and fallback to the "software implementation" (which can then hit the inliner or be a regular call).

In the case its handled up front (which is essentially any non-failure case), we shouldn't need a compile time or runtime generic instantiation because it's effectively being treated as if the user had just done the correct IL directly inline. In the case that we fallback, then the inliner checks and/or the regular processing that will happen when the call itself is compiled should handle the creation of the generic instantiations.

So my thinking is that we should be able to just slightly change when the generic instantiation is created to not include intrinsics at the very least. Not including inlined methods would be slightly more complex, particular for non-value types, but should also be possible since we know specialization is happening.

Reply:
The first thing that the token resolution does is to resolve the token into persisted instantiated `MethodDesc*`. This happens as part of the `resolveToken` JIT/EE interface call, `resolveToken` returns this instantiated `MethodDesc*` as `CORINFO_METHOD_HANDLE` and everything else after that point (`getCallInfo` to fill in `CORINFO_CALL_INFO*`, etc.) operates on that `CORINFO_METHOD_HANDLE`.

In order to avoid creating the persisted instantiated `MethodDesc*`, we would need to change the token resolution to return some kind of transient representation of the method and the instantiation, and then teach places that operate on `CORINFO_METHOD_HANDLE` to operate on this alternative transient representation instead. There would be details to decide like:
- whether the transient representation should be returned all the time or some of the time (I think it would be best to return it all the time)
- whether the JIT would need to be aware of the two different method representations or whether it would be handled transparently on the EE side of JIT/EE interface (I think it would be best to handle it transparently on the EE side)
- how to deal with type loader caches. The type loader is caching the lookups and these caches use the persistent instantiated MethodDesc today. We may need to introduce separate caches of the transient representations to avoid regressing JIT throughput.

Reply:
Thanks for the explanation! Just one more thought/question on my end...

IIRC, we already have the concept of non-instantiated descriptors as that's the difference between `typeof(List<>)` and `typeof(List<int>)` for example. You can then make the former into the latter via `MakeGenericType`. Is this somehow different from what you're referring to?

If its not different, I imagine we could effectively keep `resolveToken` "as is" for the majority of cases, but also expose a version that gets the `uninstantiated descriptor` (that is, effectively a `methodof(Unsafe.As<,>)`). This would be enough for `CEE_CALL:` handling to determine basics like "is this intrinsic", "how much IL does it have for the purposes of inlining", etc. In the case that we determine it needs further handling (for intrinsic importation or additional inlining checks), we'd then need the class handles for the generic type arguments (which are in their own additional cache separately anyways).

For the intrinsic and inlining case, we then don't create the actual generic instantiation unnecessarily. It would instead only get created when the call is directly processed.

Reply:
You are describing a design where the JIT would need to be aware of the two different method representations (my second point above).

Also, I think you still need to get the generic instantiation arguments for most `Unsafe` intrinsics in the JIT. For example, you need to know the generic instantiation arguments of `Unsafe.BitCast` to import it as intrinsics.

Reply:
> You are describing a design where the JIT would need to be aware of the two different method representations (my second point above).

👍 

> Also, I think you still need to get the generic instantiation arguments for most Unsafe intrinsics in the JIT. For example, you need to know the generic instantiation arguments of Unsafe.BitCast to import it as intrinsics.

Right, that was the "In the case that we determine it needs further handling (for intrinsic importation or additional inlining checks), we'd then need the class handles for the generic type arguments"

For most intrinsics we really just need the type handles, which are often being resolved anyway. Many intrinsics even just need simple things like `size` or the `CorInfoType`, not many need the full handle.